include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,896 @@
+                required_neg_count{pst.get_count<sidb_charge_state::NEGATIVE>()},
+                required_pos_count{pst.get_count<sidb_charge_state::POSITIVE>()},
+                required_neut_count{pst.get_count<sidb_charge_state::NEUTRAL>()}
+        {}
+
+        uint64_t omit_free_witnesses_and_count_overlap() noexcept
+        {
+            witness_set overlap{};
+
+            // find free witnesses to reduce the problem, leaving only witnesses to partition that overlap
+            for (const auto& [this_set, competing_sets] :
+                 std::array{std::make_pair(std::make_pair(std::ref(negative_witnesses), std::ref(required_neg_count)),
+                                           std::make_pair(std::ref(positive_witnesses), std::ref(neutral_witnesses))),
+                            std::make_pair(std::make_pair(std::ref(positive_witnesses), std::ref(required_pos_count)),
+                                           std::make_pair(std::ref(negative_witnesses), std::ref(neutral_witnesses))),
+                            std::make_pair(std::make_pair(std::ref(neutral_witnesses), std::ref(required_neut_count)),
+                                           std::make_pair(std::ref(negative_witnesses), std::ref(positive_witnesses)))})
+            {
+                for (witness_set::const_iterator it = this_set.first.cbegin(); it != this_set.first.cend();)
+                {
+                    if (competing_sets.first.find(*it) == competing_sets.first.cend() &&
+                        competing_sets.second.find(*it) == competing_sets.second.cend())
+                    {
+                        it = this_set.first.erase(it);
+                        this_set.second--;
+                        continue;
+                    }
+
+                    overlap.emplace(*it);
+
+                    ++it;
+                }
+            }
+
+            return overlap.size();
+        }
+    };
+
+    template <sidb_charge_state current_fill_cs>
+    static bool find_valid_witness_partitioning(witness_partitioning_state& st,
+                                                const uint64_t              num_witnesses_for_current_cs) noexcept
+    {
+        if constexpr (current_fill_cs == sidb_charge_state::NEUTRAL)
+        {
+            return st.neutral_witnesses.size() >= num_witnesses_for_current_cs;
+        }
+        else if constexpr (current_fill_cs == sidb_charge_state::POSITIVE)
+        {
+            if (num_witnesses_for_current_cs == 0)
+            {
+                return find_valid_witness_partitioning<sidb_charge_state::NEUTRAL>(st, st.required_neut_count);
+            }
+
+            for (typename witness_partitioning_state::witness_set::const_iterator it = st.positive_witnesses.cbegin();
+                 it != st.positive_witnesses.cend();)
+            {
+                const uint64_t take_witness = *st.positive_witnesses.cbegin();
```

Comment:
```suggestion
                const auto take_witness = *st.positive_witnesses.cbegin();
```

Reply:
Why so? it is not an iterator (I know Clang-Tidy says to use `auto` for those)

Reply:
Ah, really, does Clang-Tidy say that? I find it better to use ``auto`` here, because the type can be inferred from ``positive_witnesses``. So there is no need to specify it again. It also makes the code easier to maintain in general.

Reply:
I could agree with that, though just changing this one would make an anomaly wrt. the rest of the style in the code, where I use `auto` only where Clang-Tidy suggested it, to to use structured bindings. There are loads of other places where I then should use `auto` too. I'm not a fan of hidden type information for the programmer, and similarly so for reading code with hidden type information